### ISSUE TITLE
Parallel installs using git sporadically fail

### ISSUE THREAD

**Issue by mdavidn** [bundler/bundler#5552]
Parallel `bundle install` processes on the same `Gemfile` will frequently (but not always) fail due to contention on a Git lock file. I noticed this because my CI environment starts parallel builds after every update to my project.

It appears that multiple processes running `GitProxy#copy_to` will both conclude `$destination/.git` doesn't exist before running `git reset --hard` in parallel on the same clone.

I expect parallel `bundle install` processes to wait for the first to finish with a shared resource. Alternatively, `GitProxy` could clone into a process-local temporary directory.

Here's a reproduction script and the error message:

https://gist.github.com/mdavidn/66027b9727f536f9f058108560e6b88f

```
Fetching git@github.com:bundler/bundler.git
fatal: Unable to create '/Users/xxxx/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/bundler-ca5710e41dac/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
Git error: command `git reset --hard ca5710e41dacefce8951f4992378d95b470882dc`
in directory
/Users/xxxx/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/bundler-ca5710e41dac
has failed.
If this error persists you could try removing the cache directory
'/Users/xxxx/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/cache/bundler/git/bundler-f625eabf2e6cea03567de55cfa13168adc7de8ed'
```

**Comment by colby-swandale** [comment]
@indirect @segiddins any thoughts on this?

**Comment by segiddins** [comment]
Haven't had a chance to look into it yet

**Comment by segiddins** [comment]
I'm sure this issue exists, seems like we might want to try using file locks on POSIX systems to serialize git operations? Just a thought, but I think someone else would have to try and implement it, I don't see myself having the time in the near future.

**Comment by mikegee** [comment]
I think this issue can be closed because #5852 was merged.
